pycbc/results/pygrb_postprocessing_utils.py

```diff
@@ -0,0 +1,905 @@
+    signal based vetoes.  The (default) signal based vetoes are:
+    * Coherent SNR < 6
+    * Bank chi-squared reduced (new) SNR < 6
+    * Auto veto reduced (new) SNR < 6
+    * Single-detector SNR (from two most sensitive IFOs) < 4
+    * Null SNR (CoincSNR^2 - CohSNR^2)^(1/2) < null_thresh
+
+    Returns Numpy array of BestNR values.
+    """
+
+    if not trigs:
+        return numpy.array([])
+
+    # Grab sky position and timing
+    ra = trigs.get_column('ra')
+    dec = trigs.get_column('dec')
+    time = trigs.get_end()
+
+    # Initialize BestNRs
+    snr = trigs.get_column('snr')
+    bestnr = numpy.ones(len(snr))
+
+    # Coherent SNR cut
+    bestnr[numpy.asarray(snr) < snr_threshold] = 0
+
+    # Bank and auto chi-squared cuts
+    if not chisq_threshold:
+        chisq_threshold = snr_threshold
+    for chisq in ['bank_chisq', 'cont_chisq']:
+        bestnr[numpy.asarray(trigs.get_new_snr(index=q, nhigh=n,
+                                               column=chisq))
+               < chisq_threshold] = 0
+
+    # Define IFOs for sngl cut
+    ifos = list(map(str, trigs[0].get_ifos()))
+
+    # Single detector SNR cut
+    sens = {}
+    sigmasqs = trigs.get_sigmasqs()
+    ifo_snr = dict((ifo, trigs.get_sngl_snr(ifo)) for ifo in ifos)
+    for ifo in ifos:
+        antenna = Detector(ifo)
+        sens[ifo] = sigmasqs[ifo] * get_antenna_responses(antenna, ra,
+                                                          dec, time)
+    # Apply this cut only if there is more than 1 IFO
+    if len(ifos) > 1:
+        for i_trig, _ in enumerate(trigs):
+            # Apply only to triggers that were not already cut previously
+            if bestnr[i_trig] != 0:
+                ifos.sort(key=lambda ifo, j=i_trig: sens[ifo][j], reverse=True)
+                if (ifo_snr[ifos[0]][i_trig] < sngl_snr_threshold or
+                        ifo_snr[ifos[1]][i_trig] < sngl_snr_threshold):
+                    bestnr[i_trig] = 0
+    for i_trig, trig in enumerate(trigs):
+        # Get chisq reduced (new) SNR for triggers that were not cut so far
+        # NOTE: .get_bestnr is in glue.ligolw.lsctables.MultiInspiralTable
+        if bestnr[i_trig] != 0:
+            bestnr[i_trig] = trig.get_bestnr(index=q, nhigh=n,
+                                             null_snr_threshold=null_thresh[0],
+                                             null_grad_thresh=null_grad_thresh,
+                                             null_grad_val=null_grad_val)
+            # If we got this far and the bestNR is non-zero, verify that chisq
+            # was actually calculated for the trigger, otherwise print
+            # debugging info
+            if bestnr[i_trig] != 0 and trig.chisq == 0:
+                err_msg = "Chisq not calculated for trigger with end time "
+                err_msg += f"{trig.get_end()} and SNR {trig.snr}."
+                logging.error(err_msg)
+                raise RuntimeError(err_msg)
```

Comment:
Is the exception here intended? The comment claims this should just print debugging info, so maybe remove the exception and turn `logging.error()` into `logging.info()`?